pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/PulsarOffsetBackingStore.java

```diff
@@ -126,10 +127,13 @@ private void readNext(CompletableFuture<Void> endFuture) {
     }
 
     void processMessage(Message<byte[]> message) {
-        synchronized (data) {
+        if (message.getKey() != null) {
             data.put(
                 ByteBuffer.wrap(message.getKey().getBytes(UTF_8)),
                 ByteBuffer.wrap(message.getValue()));
+        } else {
+            log.debug("Got message without key from the offset storage topic, skip it. message value: {}",
```

Comment:
For offset topic, the key must not be `null`, which is guaranteed by Kafka itself. Kafka's coordinator writes messages that must have keys to the offset topic. In addition, Kafka broker prevents any write to the offset topic by a Kafka client.

Therefore, I think there is no need to perform null check here. If you insisted to check null keys, the logs here should be error because it's an unexpected case.

Reply:
It looks like the topic is produced by Pulsar producer. See `set` method:

```java
            producer.newMessage()
                .key(new String(keyBytes, UTF_8))
```

So I think the key could never be null.

Reply:
The messages managed by the backing store service itself do have the guarantees, but we do not have the guarantees from the user themselves, if the user send test messages to the offset topic (i.e. use `pulsar-client produce`) and it could cause the unexpected issues. With this scenario, the unexpected messages should be ignored and not affect the backing store, and this is why it been log as debug.

Reply:
I see.

BTW, I think we should introduce a mechanism to prevent some internal topics (like system topics) from client side in future.